### PR TITLE
Created tool to merge language arrays and updated swedish language file.

### PR DESCRIPTION
**Pull request by stenebo**
Refering to my post in the forum:
http://forum.kimai.org/index.php?topic=703.0

### CODE REVIEW

**Comment by claudegel** [comment]
Hi, Good job. I've tryed it for the fr language and I get the following message
Key array have 447 items
Value array have 436 items
Target array received 447 items
0 translations are reused, 0 values need review in file: /var/www/html/kimai-quick-note/kimai/core/language/fr-new.php
What is the difference between  the 447 items and the 436 items. Does that mean I have some missing translation or empty values?

**Comment by stenebo** [comment]
Oh, that was a quick response! I think the total count includes all key-value pairs + one for each subarray.
The difference means that the target (new) file should have received 11 new items from the first (source/key) file compared with the second (old translated values) file.
What was your first (source/key) file? en.php I assume. My test gave following result (files from repository):
php mergeLangArr.php ../core/language/en.php ../core/language/fr.php ../core/language/fr_20130801T170000.php
Key array have 447 items
Value array have 403 items
Target array received 447 items
398 translations are reused, 33 values need review in file: ../core/language/fr_20130801T170000.php

I put that new test file in my forked repo in case anyone want it.

**Comment by stenebo** [comment]
Since most people are probably using the english language file (en.php) as their starting point I think I should mention that the german file (de.php) currently contain five additional key-value pairs. 
In order to get a complete file I've used the german file as source file in a second run to get the last five lines for the swedish file.

**Comment by claudegel** [comment]
Le Thu, 01 Aug 2013 08:08:36 -0700,
Per Stenebo notifications@github.com a écrit :

> Oh, that was a quick response! I think the total count includes all
> key-value pairs + one for each subarray. The difference means that
> the target (new) file should have received 11 new items from the
> first (source/key) file compared with the second (old translated
> values) file. What was your first (source/key) file? en.php I assume.
> My test gave following result (files from repository): php
> mergeLangArr.php ../core/language/en.php ../core/language/fr.php ../core/language/fr_20130801T170000.php
> Key array have 447 items Value array have 403 items Target array
> received 447 items 398 translations are reused, 33 values need review
> in file: ../core/language/fr_20130801T170000.php
> 
> I put that new test file in my forked repo in case anyone want it.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/kimai/kimai/pull/222#issuecomment-21942905

Thanks.

first I did the
php
mergeLangArr.php ../core/language/en.php ../core/language/fr.php ../core/language/fr_new.php
 and I made all the changes encountered in the fr_new file into the
 fr.php.

then I ran the program again to see if I had miss something and get the
message I've sent to you.

so that mean all missing translation have been done ? How can I find
these 11 new items

I've made the changes in the fr.php and send it to a pull request.

very nice tool to keep us up to date for translation.

one note: your program name is mergeLangArr.php but in the pull request
sources it is called compLangArr.php in the first few lines

Thanks

## 

Claude Gélinas agr.
Phyto Ressources Inc

**Comment by claudegel** [comment]
Le Thu, 01 Aug 2013 08:23:05 -0700,
Per Stenebo notifications@github.com a écrit :

> Since most people are probably using the english language file
> (en.php) as their starting point I think I should mention that the
> german file (de.php) currently contain five additional key-value
> pairs. In order to get a complete file I've used the german file as
> source file in a second run to get the last five lines for the
> swedish file.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/kimai/kimai/pull/222#issuecomment-21943987

if I compare the de.php with my updated fr.php I get 

Key array have 451 items
Value array have 403 items
Target array received 451 items
0 translations are reused, 0 values need review in file

does that mean my fr.php is up to date ? I don't understand the
difference between 451 and 403

do I need to do something or is it just as the orderring of the array

## are different in fr.php and fr_new.php

Claude Gélinas agr.
Phyto Ressources Inc

**Comment by stenebo** [comment]
I'm a bit puzzled, I can't reproduce "0 translations are reused, 0 values need review in file".
Could you please send me the three files who gave that result? I'll send you a email so you get my address.
Please also write the command so I know the file order.

Tanks for your note about the old file name compLangArr.php, I'll correct it.

I will also try to improve the file feedback to clarify things.
Really great with this feedback from you, thanks!
/Per

**Comment by kevinpapst** [comment]
Guys, I did not checkout your script, but have you seen: 
https://github.com/kimai/kimai/blob/master/core/shellscripts/find_missing_translations.php
This exists there for quite a long time ;-) Please incorporate your changes to that script instead of creating a new solution.

**Comment by stenebo** [comment]
Hi Kevin! I tested your script and I have a couple of notes:
1. We have slightly different aim even though the goal are the same:

find_missing_translations.php: Simply show missing keys in all lang files, but only in the base array, not in subarrays.

mergeLangArr.php: Merge the array structure from one file (eg. de.php or en.php) with already translated values from another (old) lang file and creating a new file with the resulting array.
The purpouse is to minimize the work needed to update an existing but outdated lang file and remove the need for error-prone manual editing of array structure.
1. find_missing_translations.php could easily be improved to also print missing keys in subarrays. Do you want me to do that?
2. If you still want the scripts to be merged, what should it do?
3. If you want to keep mergeLangArr.php I suggest that we move it from /test to /core/shellscripts
4. I'm a git newbie so forgive me for surplus files in my pull request.
   /Per

**Comment by claudegel** [comment]
Hi, I've tested both and would prefer the mergeLangArr.php as it add in the language file the missing translation with a comment at the end of the line -->REVIEW so it is easy to fix the missing translations.

The only problem is that it look like mergeLangArr.php do not catch all missing translations.

ex when I compare DE with FR  find_missing_translations.php give one missing array and mergeLangArr.php say everything is uptodate. In fact the 'errorMessages' => array() is completely missing from FR translation

So I think we need to merge both into one perfect tool that add automatically the missing translation into each file with a note to show what need to be translated

**Comment by stenebo** [comment]
I've just tested with the following result (with kimai repo lang files and latest mergeLangArr.php from my repo):
php mergeLangArr.php ../core/language/de.php ../core/language/fr.php ../core/language/fr_new.php
397 old values are reused, 38 new values need to be reviewed in file: ../core/language/fr_new.php

The subarray 'errorMessages' => array(... are present in the new file even as it was missing in the old fr.php.

When I run it again for verification with the resulting file from previous run:
php mergeLangArr.php ../core/language/de.php ../core/language/fr_new.php ../core/language/fr_verified.php
435 old values are reused, 0 new values need to be reviewed in file: ../core/language/fr_verified.php

The same test with find_missing_translations.php (slightly modified for this test to only check fr_new.php):
No translation missing!

In summary: I could not reproduce a missing subarray. If the error still exist I need more help to pinpoint it.
/Per

**Comment by kevinpapst** [comment]
If your tools finds at least the same amount of keys than mine does, then I am fine if you replace mine completely.

We just should not keep two files with the same purpose in the repo. So either move it to shellscripts/fix_missing_translations.php or even overwrite my file.

For your next pull request: please do not include multiple topics in one branch (here: a new tool mergeLangArr.php  and the updated translation sv.php). Translations can (for example) always be merged directly, where code changes should not be included until the feature freeze for 0.9.3 is over. 

And thanks for your effort :+1:

**Comment by claudegel** [comment]
Ok I've sent a fix to Per for the missing key so now both script are findind the same amount and mergeLangArr.php is doing all the dirty job for me :)

**Comment by stenebo** [comment]
I've reverted all changes and are trying to do it right this time ;-)
Overhauled lang tool committed that replaces current file and nothing else.
I'll send the updated lang file sv.php later in another request.
/Per
